AEP/aep-00004.md

```diff
@@ -0,0 +1,163 @@
+
+# Motivation
+The motivation of this AEP is to proactively treat a previsible bottleneck on the database. 
+
+Currently in ArmoniK, MongoDB is deployed using a simple replica set architecture, meaning that the database is composed of multiple identical instances. However, only one instance of the database supports write operations (the primary node), resulting in a potential bottleneck on this node. Additionnaly replicating the whole database is not very efficient for read operations, as the DB engine has to parse bigger datasets.
+
+Finally, the current MongoDB module in our infrastructure's code is defined kind of manually with multiple resources being defined, which is not  necessarily the best choice as deploying a database, especially a sharded one, is a complex task. 
+
+# Rationale
+Sharding the database would alleviate the tension on all the nodes. For the primary nodes, the write operations would be distributed among other shards' primary node, and for the secondary nodes there can be some situations where only the right shard is queried (it will be explicited below in this AEP) and as mentionned above, the read operations are faster on a sharded cluster since the data is distributed among multiple instances.
+
+This sharded (or not) MongoDB deployment would be defined in our infrastructure code as a Helm release resource (called by Terraform). A Helm release is an instance of a Helm chart. Helm charts are templates that allow to deploy resources within a Kubernetes cluster in configurable way.
+Using Helm charts conveys the complexity of defining a deployment towards the authors of a Helm chart who had more time to think about how MongoDB should be deployed within a Kubernetes cluster.
+
+Instead of  manually defining a MongoDB deployment with Terraform modules as it is currently done with ArmoniK's infrastructure, Helm charts do it for us. All we have to do is to define the right configuration for the MongoDB deployment.
+
+During our reflexion, some other alternatives to responding this issue were considered :
+- First, adding some resources and modifying some others in our Terraform module. But according to the KISS and "don't reinvent the wheel" principles, we found this solution unrelevant as others developers are able to provide us out-of-the-box MongoDB deployments, which is often a better way of building softwares as the complexity is shifted to the Helm chart definiton.
+- Using Kubernetes operator for MongoDB, but as far as we have explored this solution, we were not able to find an operator that supports sharding.
+
+To conclude, using a MongoDB Helm chart would allow ArmoniK to benefit from an out-of-the-box MongoDB deployment to separate the complexity of it from the sheer infrastructure code defined in Terraform, as well as benefit from the maintenance of the Helm chart used, again in the scope of dividing complexity.
```

Comment:
```suggestion
To conclude, using a MongoDB Helm chart would allow ArmoniK to benefit from an out-of-the-box MongoDB deployment to separate the complexity of it from the sheer infrastructure code defined in Terraform, as well as benefit from the maintenance of the Helm chart provided by the MongoDB community itself.
```